### PR TITLE
Adds gem i18n tasks and creates build node to run tasks

### LINKED ISSUE #2626: Avoid missing i18n translations
## Description

There doesn't seem to be an easy way right now to make sure that translation keys we use indeed have corresponding entries in the main language file.

There's a chance to miss it if a developer introduces a new translation key but forgets to add it to the language file.

And if a translation is missing, the non-underscored version of the key is used, so a missing "new_feature" appears as "New Feature" (wrapped in `span.translation_missing`). This translation often still makes sense for the `development` language `en`, so you would think everything is okay when manually testing in the browser.

These are small enough changes that can help us avoid these missing translations:

* For translations used in the Ruby code, in automated tests (only `test` environment), raise error when encountering a missing translation key. We have good test coverage, so I think this would be very helpful.
* For translations used in the JS, warn in the browser console if a translation is missing.
* Make it more obvious in the UI if a certain text is only from the non-underscored version of the translation key. As mentioned, these are wrapped in `span.translation_missing` so can easily be styled.

## Expected Behavior

Warnings when using a translation key that is missing from the language file

## Actual Behavior

No changes in actual behaviour if translations are okay

## Steps to Reproduce

1. Add `t("missing_key")` to a page, and check page. The page would not complain, and render this as "Missing Key".
2. The following test would not complain:
```ruby
require 'spec_helper'

describe "Missing translation" do
  it "raises error when a translation is missing" do
    I18n.t("missing_key")
  end
end
```

## Context

Lack of confidence when adding/updating translation keys

### PR DESCRIPTION
**Pull request by filipefurtad0**
#### What? Why?

- Closes #2626

This gem looks like a good toolkit to maintain our translations.
It can run as a spec, and I've created a build node so it can run separately from other checks.

There are some things which need addressing before it can turn green, but these are probably worthwhile. My proposal would be to set the respective examples as pending (on `spec/i18n-tasks/i18n_spec.rb`), and fix them as we go.

Edit: we've discussed this on in the past [here](https://openfoodnetwork.slack.com/archives/C02TZ6X00/p1677535939854909?thread_ts=1658137192.453269&cid=C02TZ6X00).

#### What should we test?

- green build (after setting failing examples as pending)

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -502,3 +502,50 @@ jobs:
...
+
+      - name: Set up database
+        run: |
+          bundle exec rake db:create
+          bundle exec rake db:schema:load
+      - name: specs on i18n-exceptions
+        run: bundle exec rspec ./spec/i18n-tasks/i18n_spec.rb
+        run: i18n-tasks missing
+        run: i18n-tasks unused
+        run: i18n-tasks check-consistent-interpolations
```
Comment (sigmundpetersen) [inline .github/workflows/build.yml:551]: The build is complaining, it seems you cannot have several `run` statements after each other

Comment (filipefurtad0) [reply .github/workflows/build.yml]: Thanks @sigmundpetersen :-D

Comment (filipefurtad0) [reply .github/workflows/build.yml]: I was trying to find a way to output the list of errors as well... So one could see immediately what the issue is, without having to run it locally.

Comment (filipefurtad0) [reply .github/workflows/build.yml]: It seems the syntax the pipe is used above in the file, but still no output. I'll just remove it for now.

**Review by rioug — APPROVED** [review]
Nice :clap: , I like the approach. I would just note the following : 

> By default, dynamic keys such as t "cats.#{cat}.name" are not recognized. I encourage you to mark these with [i18n-tasks-use hints](https://github.com/glebm/i18n-tasks#fine-tuning).

I would expect we don't have many of them so it should be fine.
Is whatever needs to be done before we can turn this on tracked somewhere ?

**Review by dacook — APPROVED** [review]
Nice, so as the gem readme says, it has two benefits that are relevant to us:

1. **Report keys that are missing**
    Your [other PR](https://github.com/openfoodfoundation/openfoodnetwork/pull/11906) seeks to solve that by detecting missing keys in specs. I thought that would be enough.
    But it's possible that we have un-tested code, so this code analysis tool could help with that. In fact, if it picks up a missing translation in un-tested code, that would remind us to add a test. 

2. **Report keys that are unused**
    The en.yml file is huge, so it would be great to ensure we dont' have unnecessary keys!

It's surprising to see it took 5 minutes for the task to run. Perhaps we can configure it to optimise if needed.

**Comment by dacook** [comment]
And as Gaetan mentions, we should either fix the pending translations, or ensure they are tracked in issues.

@filipefurtad0 I think some analysis of the results will be necessary, because the numbers are huge!  maybe it needs configuring better. Would you be able to review these before merging?


>  Missing 22591 i18n keys, run `i18n-tasks missing' to show them
> 
>  48358 unused i18n keys, run `i18n-tasks unused' to show them
> 
>  The following files need to be normalized:
>          config/locales/en.yml
>          config/locales/ar.yml
>          ...(all of them)
> 
> 59 i18n keys have inconsistent     interpolations.

https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/7211174521/job/19645993874?pr=11938

**Review by mkllnk — COMMENTED** [review]
Nice idea. The resulting numbers seem too high though.

I'm still a bit cautious about this gem. Dynamic keys are used and maintaining exceptions may become a lot of work for not much benefit.

> Report keys that are missing

This is important. We already cover that as long as the code is covered by specs. Untested code is a problem but I haven't seen good examples yet how this gem shows us what's missing. And I would strongly advocate for better code coverage which is more important and more reliable than this gem.

> Report keys that are unused

Nice to have. Our file is big but I don't think that it affects performance much. It will also be always some kind of big because we have a lot of text in our application. We may be trading more efficient editing of en.yml for maintenance of this gem and configuration.

I'm sorry if I'm too negative here. I just wanted to be very clear about cost and benefit of new tools.

**Comment by filipefurtad0** [comment]
Thanks everyone. Agree this needs some work before merging, at least to interpret the results. I'll try to address the 4 examples in `./spec/i18n-tasks/i18n_spec.rb`.

1) `Missing translations (19574)` - I think this does not necessarily mean, we need code changes. I think it might just mean, that there are keys which require translation. For example (notice the red lettering):

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/9ae9fec5-a572-4278-9b55-224a3c4c7ce1)

This is the [xit example in product_import_spec.rb](https://github.com/openfoodfoundation/openfoodnetwork/blob/97985dbe8f118c2a80dae34cd727c6c380a01b6f/spec/system/admin/product_import_spec.rb#L757).

This is the key `admin.product_import.model.category_not_found` in the `:es` locale. If we run:
`$ i18n-tasks missing | grep category_not_found`

we get all the locales for which this key is missing:

```
Missing translations (19574) | i18n-tasks v1.0.13
| de_CH  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   el   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_AU  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_BE  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_CH  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_DE  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_IN  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_NZ  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_PH  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_US  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| en_ZA  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   es   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| es_CO  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| es_CR  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| es_US  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| fil_PH | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| fr_CH  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| fr_CM  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   hu   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| it_CH  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   ko   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| nl_BE  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   pl   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   pt   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| pt_BR  | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   sv   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   tr   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|   uk   | admin.product_import.model.category_not_found                                                                              | en doesn't match allowed categories. See the correct categories to choose from on the product import page, or check that there's no misspelling.  
```                                                              

How do we deal with this? I guess this does not relate to the code, it's rather a missing entry on the each respective locale... And looking this example, it does make the number `Missing translations (19574)` more plausible.

So, I guess good to know, but I'd say this is not related to the code itself.

2. This points to a lot of keys indeed... WIP.

To be continued...


One question though:

> Untested code is a problem but I haven't seen good examples yet how this gem shows us what's missing.

Can we trigger missing translation errors in tests which are not view or system specs? As I understood it, this PR https://github.com/openfoodfoundation/openfoodnetwork/pull/11906 only looks at these two types.

**Comment by mkllnk** [comment]
I think that the error raising is valid for all specs, not just system and view.

Can we limit this gem to en.yml and ignore the other locales? It's the responsibility of Transifex and the translators to fill that in.

**Comment by filipefurtad0** [comment]
Ok, I've set the gem to work only on `en` locale. It makes in easier to get some value from the gem. However, the suggested commands end up breaking our en.yml file:

- if we run `i18n-tasks remove-unused` then we get a red build - around 1000 keys are removed. The first key which seems to be missing is `en.spree.errors.messages.invalid_instagram_url`; there might be others...
- running `i18n-tasks missing` indicates 50 keys as not having correspondence in the en.yml file. I'm not entirely sure these are not false positives...
- running `i18n-tasks normalize` trims the file from comments and sets up a different indentation... I'm not sure this is useful.

**Summary**
I don't know. The fact that it breaks the build with a conservative use (after setup to affect only `en.yml`) is not particularly nice... Maybe it would indeed require much more effort than antecipated, which seems to be opposite to https://github.com/openfoodfoundation/openfoodnetwork/pull/11906 (which rather looks like a quick win).

The only thing I still think could be valuable is understanding whether those 50 missing translations are "real". We would not need to merge this PR for that, so I'd be up for closing it.

**Comment by mkllnk** [comment]
It may be worth looking through those 50 "missing" ones. All output of this gem has to be manually reviewed because it's only guessing.

**Comment by filipefurtad0** [comment]
> It may be worth looking through those 50 "missing" ones.

I'll move this to in dev and convert it to draft, for now (red build). This way we can review that output without mistakenly merging it.

**Comment by filipefurtad0** [comment]
Ok, I've checked the entries for :en locale on https://github.com/openfoodfoundation/openfoodnetwork/pull/11958, I think we can close this PR. Thanks for all the help and discussion around this!

I think the summary is:

This gem is great, but generates too many false positives, namely due to dymamic key generation. As such, we cannot use it on our CI. Instead, we can run it from time to time, to make sure we're not introducing new keys.

Closing.

**Pull request closed without merging**